### ISSUE TITLE
Configurable keyboard shortcuts

### ISSUE THREAD

**Issue by Ketan2010** [shaka-project/shaka-player#3223]
**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
YES, there is no duplicate open issue for this feature.

**Is your feature request related to a problem? Please describe.**
No, Here is suggestion for better usability and accessibility.

**Describe the solution you'd like**
For better usability there should be provision to perform actions on keyboard key's input. 
for example,
1. f key => Full screen
2. Space button => To start and stop video
3. p key => Picture in picture
4. m key => to mute and unmute audio

This feature helps everyone suffering from any type of disability, especially those with limited motor movements. 
I would like to work on this feature as I am looking forward to contribute in this project through Gsoc'21.
Thanks

**Comment by nirnayjain** [comment]
I would like to work on this issue

**Comment by nbcl** [comment]
This sounds like a great feature.

I would like to suggest that you also consider `c` to toggle closed captions if available!

**Comment by Ketan2010** [comment]
Hello @nbcl Sir,
Thanks for valuable suggestion , I will surely look into it.

**Comment by sradhanjan** [comment]
@nbcl So we can work on the projects outside of GSOC idea list right?

**Comment by nbcl** [comment]
Hello @sradhanjan!

Here's a reply from @michellezhuogg on 2607 that I think could be helpful.

> You'll need to go through the application process and officially accept the applications, to get recognized as a GSoC participant, and also get our guaranteed support and mentorship in summer.

> On the other hand, we welcome contributions from developers all the time. If you're interested in any task and are confident to come up with a concrete solution without our help, please send us your proposal (for a project) or your PR (for a simple task). 

Hope this helps!

**Comment by joeyparrish** [comment]
Hi everyone,

Any key shortcuts like this need to be configurable.  Existing Shaka-based applications would not like to be surprised by key bindings that may conflict with those they already have at the app level.  Especially if you want something easy to do by mistake (like a single keystroke) to do something big (like toggle captions or go fullscreen).

Thanks!

**Comment by Ketan2010** [comment]
thanks @joeyparrish sir,
I would like to work on full screen toggle on pressing `f` key as I am looking forward to contribution for Gsoc'21.
I need your initial guide regarding implementation.  
Thanks.

**Comment by michellezhuogg** [comment]
@Ketan2010 ,

Thank you for your idea! We'll need add a configuration field for the key shortcuts, as part of the [UI configuration](https://shaka-player-demo.appspot.com/docs/api/shaka.extern.html#.UIConfiguration).
That way, the developers can configure:
- whether to enable the shortcut keys
- which keys they want to bind the behaviors with.

For example, the developer can enable entering full screen with the "f" key with:
```
const config = {
  keyShortcuts {
    key: 'f',
    control: 'fullScreen'
  }
}
```

The implementation steps would be:
1. Add a configuration field
2. Read the configuration values, and enable controls with the key shortcuts
3. Add tests
4. Add docs

If you need closer guidance apart from the [tutorial](https://shaka-player-demo.appspot.com/docs/api/tutorial-welcome.html) and [UI config tutorial](https://shaka-player-demo.appspot.com/docs/api/tutorial-ui-customization.html). , we'll provide official mentorship once the GSoC program starts.

**Comment by surajkumar-sk** [comment]
@Ketan2010 @michellezhuogg Is this an idea for GSOC or just an enhancement idea on which anyone can work now . I would like to implement this feature for getting myself more familiar with the code base of shaka player, If @Ketan2010  doesn't have a problem with that or else I'll just leave this as a GSOC idea .

**Comment by Kotekczu** [comment]
To be able to use the keyboard shortcuts for forwarding and backwarding it's needed to click first on a seek bar point. Is this intended?

**Comment by michellezhuogg** [comment]
> To be able to use the keyboard shortcuts for forwarding and backwarding it's needed to click first on a seek bar point. Is this intended?

Yes.

**Comment by Ketan2010** [comment]
> @Ketan2010 @michellezhuogg Is this an idea for GSOC or just an enhancement idea on which anyone can work now . I would like to implement this feature for getting myself more familiar with the code base of shaka player, If @Ketan2010 doesn't have a problem with that or else I'll just leave this as a GSOC idea .

Hello @surajkumar-sk,
This idea is not listed for Gsoc by shaka player. I am just giving my thoughts to contribute and familiarize with repository structure and code. I already started working on this feature but you also can try. you can see actual ideas for Gsoc'21 by shaka player [here](https://nightly-dot-shaka-player-demo.appspot.com/docs/api/tutorial-gsoc-ideas-list.html).

**Comment by surajkumar-sk** [comment]
@Ketan2010  , i meant were you considering this idea for GSOC . it's your idea so it's better you work on it .

Anyways , have fun implementing it , I'll be looking forward for the PR . If there's any problem feel free to ask .

**Comment by Ketan2010** [comment]
Hello @michellezhuogg @joeyparrish Sir,
I seen that you have added `Gsoc ` label to issues/suggestions issued by Gsoc contributors. If so, can you please add the same for this issue? if it is really going to beneficial.
Thanks

**Comment by Ketan2010** [comment]
Hello,
Current keyboard shortcuts(right and left key for forward and backward) can be accessible only if we click on seek bar once. is it issue or any specific intension behind it ?
I am trying to implement configurable keyboard shortcuts. so, that should behave same as existing one or there should not be need to click on controls buttons to use that shortcut?

I think there should not be any need to click to use shortcut but I just want to know your opinion.

I tried it by adding new case in switch-case of `onKeyUp_` function in `control.js` file as follow,
 ```
case 'f':
        if (this.config_.enableFullscreenOnfPress) {
          this.toggleFullScreen();
        }
        break;
```

where enableFullscreenOnfPress can be configure as follow,
```
const config = {
    enableFullscreenOnfPress: true
  };
ui.configure(config);
```

This implementation is working fine but it needs to click first on any control button to use f key shortcut.
of course I will try to implement it as suggest by  @michellezhuogg so that developers can configure:

whether to enable the shortcut keys
which keys they want to bind the behaviors with.

But this is just start and I want to know whether I am going right or no.
Thanks!

**Comment by surajkumar-sk** [comment]
@Ketan2010 Yes, that feature is intended, shaka-player is used by many users to play videos online for their website. Adding features or buttons or controls which are not implemented in official Shaka-player is pretty easy by using few methods mentioned in docs. so, almost everyone uses different ways to implement the given feature or add a new feature without changing the shaka-player files. As you suggested if clicking of 'f' or any key configured by the user makes the video go fullscreen without clicking or selecting the video div then pressing left arrow keys would rewind or forward the video or increase the volume, space to stop, home to go to starting of the video all of this by default without the user having the control over it but the user might want to use those keys to trigger something else and might want the video to just play by click or be uncontrolable. if the user wants keyboard controls to work without clicking on the video he could just set `this.config_.enableKeyboardPlaybackControls='true'; ` which will enable keyboard shortcuts to run without selecting video container. but by default they are false. so when the user browsing the video uses a tab to navigate he intends to use video controls with the keyboard so `this.config_.enableKeyboardPlaybackControls` becomes true and keyboard shortcuts run. so, changes in shaka-player files should be made taking the defaults into consideration, if you want to use shortcuts without selecting the video make `this.config_.enableKeyboardPlaybackControls ` set this to true in your files . Do not commit those changes.

If there's anything you didn't understand feel free to ask.

**Comment by joeyparrish** [comment]
@surajkumar-sk wrote:
> if the user wants keyboard controls to work without clicking on the video he could just set this.config_.enableKeyboardPlaybackControls='true'; which will enable keyboard shortcuts to run without selecting video container. but by default they are false. so when the user browsing the video uses a tab to navigate he intends to use video controls with the keyboard so this.config_.enableKeyboardPlaybackControls becomes true and keyboard shortcuts run.

To clarify, since many of us are new here, though @surajkumar-sk speaks with great confidence, he is also new here and doesn't speak for the project.  He is partly right, when he says that the `enableKeyboardPlaybackControls` config must be enabled for keyboard controls to work.  However, he is mistaken to say that it defaults to `false`.  It defaults to `true` (see `ui/ui.js`).  It is also not changed by tabbing.

@Ketan2010 wrote:
> Current keyboard shortcuts(right and left key for forward and backward) can be accessible only if we click on seek bar once. is it issue or any specific intension behind it ?

That sounds like focus to me.  When you click a control, it receives focus, and keystrokes now go to that control.  You can also focus a control by using the tab key, to cycle through the various controls on a page.  If you listen for key events on the document or window, you get something like global keyboard shortcuts.  We listen for left and right on the video controls container, so if any control is in focus, those left and right keys will seek.  (See `ui/controls.js`)

**Comment by surajkumar-sk** [comment]
@joeyparrish thank you for clarifying, I just realized that enablekeyboardplaybackcontrols are true by default. i was working on that improvement with seekbar and couldn't figure out why enablekeyboardplaybackcontrols is true when its default should be false. 

just another small clarification why does enablekeyboardplaybackcotrols has a value false inside `onkeyup()` function when the video gets loaded. It tends to have a value false when inside `onkeyup()` so I assumed false was the default and when tab is used the video container gets `'shaka-keyboard-navigation` gets added to classlist making enablekeyboardplaybackcotrols true .

**Comment by surajkumar-sk** [comment]
@Ketan2010 I am sorry for posting the wrong information. I'll make sure I get my facts clear the next time I try posting something.😅

**Comment by surajkumar-sk** [comment]
@joeyparrish  A small confusion led to this big blunder, Eventmanger is listening to key-down for windows and keyups for video container. so onkeydown() gets executed globally and onkeyup() get executed only when focus is inside video container. i was thinking the onkeup() function quits in between because `enableKeyboardPlaybackControls` has default as false and return statement is read and the function ends without performing actions and when the tab is pressed video container becomes active and  `enableKeyboardPlaybackControls` becomes true. I am posting this because while understanding the code it is a very simple blunder one can make which can lead to a wrong understanding of code.

**Comment by joeyparrish** [comment]
No worries about the misunderstanding.  Often, there is some better way we could structure or comment or document the code to make it easier to understand.  Thanks for posting again to clarify what you've learned from the code!

In this case, the keydown handler is watching the window to keep track of tab navigation through the controls, while the keyup handler is watching the controls and actually implementing the keyboard controls.

I honestly do not know why we used two different events for these, though it does make sense that the tab-tracking would require listening to global events from the window.  If I had a bit more time, I would dig through the commit messages and comments to see if there were any clues to explain why we use different events for these.  It could just be an oversight or an arbitrary choice.

**Comment by Ketan2010** [comment]
> @Ketan2010 wrote:
> 
> > Current keyboard shortcuts(right and left key for forward and backward) can be accessible only if we click on seek bar once. is it issue or any specific intension behind it ?
> 
> That sounds like focus to me. When you click a control, it receives focus, and keystrokes now go to that control. You can also focus a control by using the tab key, to cycle through the various controls on a page. If you listen for key events on the document or window, you get something like global keyboard shortcuts. We listen for left and right on the video controls container, so if any control is in focus, those left and right keys will seek. (See `ui/controls.js`)

@joeyparrish Sir, 
So any newly added keyboard shortcut (Full screen, PiP, Caption) should work on **focus** phenomenon? or that should watch window to keep track as `keydown` handler?
